[BitwiseCopyable] Allow suppression via ~.

nate-chandler opened the pull request:
In addition to the existing language mechanism of `@available(*, unavailable)` on an extension.

nate-chandler commented:
@swift-ci please test

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -595,7 +595,10 @@ class alignas(1 << DeclAlignInBits) Decl : public ASTAllocated<Decl> {
-    IsComputingSemanticMembers : 1
+    IsComputingSemanticMembers : 1,
+
+    /// Whether ~BitwiseCopyable appeared in the type's inheritance list.
+    SuppressedBitwiseCopyable : 1
```
I'd prefer not to add a specific bit for BitwiseCopyable here, because there are other protocols for which the `~` will make sense now and in the near future (e.g., `Sendable`, `Equatable`, and `Hashable`). Can we instead store a set of suppressed protocols elsewhere?

nate-chandler replied:
Changed this bit to indicate whether suppressed conformances had been computed.

DougGregor commented on lib/Sema/TypeCheckRequestFunctions.cpp:
```diff
@@ -73,13 +73,29 @@ Type InheritedTypeRequest::evaluate(
+    if (auto itr = dyn_cast<InverseTypeRepr>(typeRepr)) {
+      Type inheritedTy = resolution->resolveType(itr->getConstraint());
+      auto kp = inheritedTy->getKnownProtocol();
+      if (kp == KnownProtocolKind::BitwiseCopyable) {
```
Can this be generalized to handle any known protocol (that isn't invertible, I guess), so there are fewer places to update when we want to thread through one of the others?

nate-chandler replied:
Changed to return any non-invertible known protocol and the TypeRepr (for diagnostics).

DougGregor commented on stdlib/public/core/MemoryLayout.swift:
```diff
@@ -40,10 +40,7 @@
-
-@available(*, unavailable)
-extension MemoryLayout: _BitwiseCopyable {}
+public enum MemoryLayout<T: ~Copyable>: ~_BitwiseCopyable, Copyable {}
```
Why do we care that MemoryLayout is not bitwise copyable? It has no cases anyway.

nate-chandler replied:
That was requested here: https://github.com/apple/swift-evolution/pull/2314#discussion_r1508201954 .

DougGregor commented on lib/AST/TypeCheckRequests.cpp:
```diff
@@ -116,16 +116,33 @@ bool InheritedTypeRequest::isCached() const {
+  case InheritedTypeResult::Kind::SuppressedInference: {
+    auto *decl = cast<const TypeDecl *>(std::get<0>(getStorage()));
+    auto *nominal = const_cast<NominalTypeDecl *>(cast<NominalTypeDecl>(decl));
+    nominal->setSuppressesBitwiseCopyable();
```
Let's generalize to store the `KnownProtocolKind` here.

nate-chandler replied:
Instead of storing the field here, the storage occurs at the end of `checkInheritanceClause` which allows for diagnosing duplicate suppressions.

DougGregor commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -78,13 +79,63 @@ void simple_display(
 
 void simple_display(llvm::raw_ostream &out, ASTContext *ctx);
 
+/// Emulates the following enum with associated values:
+/// enum InheritedTypeResult {
+///     case inherited(Type)
+///     case SuppressedInference(KnownProtocolKind)
+///     case `default`
+/// }
+class InheritedTypeResult {
```
This will introduce a second representation for the `~Protocol` syntax in the AST: For the "invertible" protocols like `Copyable`, we represent them as a `ProtocolCompositionType`, which has its own list of inverses. I think that representation doesn't make sense for this use case, so it's okay to have two.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[NFC] AST: Define "repressible" known protocol.

This is the subset of suppressible protocols that are not invertible.
This commit doesn't add any such protocols.

nate-chandler committed:
[Sema] Enable suppression of inferred conformances.

Add the machinery to support suppression of inference of conformance to
protocols that would otherwise be derived automatically.

This commit does not enable any conformances to be suppressed.

nate-chandler committed:
[NFC] stdlib: Separate type decl from members.

nate-chandler committed:
[BitwiseCopyable] Suppress via ~.

In addition to the existing language mechanism of
`@available(*, unavailable)`.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -78,13 +81,75 @@ void simple_display(
 
 void simple_display(llvm::raw_ostream &out, ASTContext *ctx);
 
+/// Emulates the following enum with associated values:
```
This seems like a lot of new code for just the new `~BitwiseCopyable` syntax. Can you instead find a way to generalize the existing logic for `~Copyable`?

nate-chandler replied:
I think having a different representation for this case--the case of a protocol that isn't conformed to by all types by default (i.e. unlike `Copyable`) but is inferred automatically for some types--makes sense.

slavapestov commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -87,13 +88,101 @@ static Type containsParameterizedProtocolType(Type inheritedTy) {
 /// file.
-static void checkInheritanceClause(
-    llvm::PointerUnion<const TypeDecl *, const ExtensionDecl *> declUnion) {
+ArrayRef<InheritedEntry> InheritanceClauseRequest::evaluate(
```
This is the wrong thing to turn into a request. `checkInheritanceClause()` only exists to emit diagnostics, and we don't want to emit the diagnostics in a secondary file. TypeCheckDeclPrimary.cpp shouldn't have logic for computing things, it's just diagnostics

nate-chandler replied:
Removed this requestification and moved `SuppressesConformanceRequest` into `TypeCheckRequestFunctions.cpp`.

slavapestov reviewed (changes requested):
I think you want to look at the callers of getDirectlyInheritedNominalTypeDecls() in ConformanceLookupTable.cpp, and put the logic there. getDirectlyInheritedNominalTypeDecls() should be generalized to collect and return these new "repressible" protocols.

nate-chandler commented:
@swift-ci please build toolchain

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test
